### ISSUE TITLE
CdkPipeline: How to use existing service roles for code pipeline actions

### ISSUE THREAD

**Issue by ColeSiegelTR** [aws/aws-cdk#15649]
### The Question

Is there a way to override the generated pipeline service roles with existing roles? CdkPipeline creates a bunch of roles below which are used by the CodePipeline:

- codepipelineBuildSynthCodePipelineActionRole
- codepipelineUpdatePipelineSelfMutateCodePipelineActionRole
- codepipelineBuildSynthCdkBuildProjectRole
- etc.

I am wondering if there is a way to override these, because I need to use a pre-existing service role created by our security team instead. These generated roles are missing the required permissions to assume the stack synthesizer roles.

Sample code:

```ts
        const pipeline = new cdkp.CdkPipeline(this, 'cdk_pipeline', {
            selfMutating: true,

            codePipeline: new codepipeline.Pipeline(this, 'code_pipeline', {
                pipelineName: pipelineConfiguration.stackName,
                role: cicdRole,
                artifactBucket: existingEncryptedBucket,
            }),

            cloudAssemblyArtifact: cloudAssemblyArtifact,
            sourceAction: sourceAction,
            synthAction: cdkp.SimpleSynthAction.standardNpmSynth({
                environment: {
                    privileged: true
                },
                subdirectory: 'iac/cdk',
                sourceArtifact: sourceOutput,
                cloudAssemblyArtifact: cloudAssemblyArtifact,
                installCommand: 'npm ci'
            })
        })
```

### Environment

  - **CDK CLI Version:** 1.110.0
  - **Module Version:** 1.110.0

Referring to the service role here:

<img width="1211" alt="Screen Shot 2021-07-19 at 12 37 07 PM" src="https://user-images.githubusercontent.com/73616389/126195615-e15eb444-0405-402d-8bb9-6fd81b7685ba.png">

**Comment by ColeSiegelTR** [comment]
Similar issue - https://github.com/aws/aws-cdk/issues/12856

**Comment by skinny85** [comment]
Re-assigning to @otaviomacedo, who knows more about CDK Pipelines than I do 🙂.

**Comment by peterwoodworth** [comment]
The interface `IAction`'s prop `ActionProperties` contains a prop called `role`. Using this prop will result in using that role instead of creating `codepipelineBlahblahblahActionRole`

https://github.com/aws/aws-cdk/blob/7e41b6b745ff185110b7aba3db15d30bb35577bc/packages/%40aws-cdk/aws-codepipeline/lib/pipeline.ts#L694-L709

However, the role has to be owned by AWS. This is the only way that the CDK won't create roles related to pipeline actions.

That's just for the ActionRoles, I haven't looked into how the other roles are created. So in the case of just action roles, you'd need to make a feature request to be able to configure your roles if the current way to do it doesn't help with your use case.

**Comment by ColeSiegelTR** [comment]
Hey @peterwoodworth, thanks for your response!

Can you clarify what you mean by `the role has to be owned by AWS`? Does it mean I cannot create the role myself in my own AWS account?

It looks like I might be able to use the `IAction` interface to modify the roles for the `SourceAction` and `SynthAction`, but it is unclear how to change other generated roles such as `codepipelineBuildSynthCdkBuildProjectRole`, any suggestions there?

**Comment by peterwoodworth** [comment]
I'm not entirely certain what `owned by AWS` exactly means in this case. What the code checks for is if the `owner` property on your [`IAction`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-codepipeline.IAction.html) is equal to `aws`. However, I don't see a way to actually configure these properties with either SimpleSynthAction or sourceAction, so I don't know how exactly this prop is naturally supposed to be filled in. The reason the owner has to be AWS is stated in the code: https://github.com/aws/aws-cdk/blob/7e41b6b745ff185110b7aba3db15d30bb35577bc/packages/%40aws-cdk/aws-codepipeline/lib/pipeline.ts#L711-L712

The other thread you linked had some info that supports my gut feeling (which is that there's not currently a way to override all the generated roles). The CDK was built so that you don't have to worry about configuring roles and policies by hand, and looking at the code it doesn't look like there's a built in way to easily use your own roles instead. 

I wonder if you could use [escape hatches](https://docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html) to manually replace all the roles. I don't recommend doing this however, as that would probably get ugly fast.

@otaviomacedo your input would be greatly appreciated here

**Comment by otaviomacedo** [comment]
The role `codepipelineBuildSynthCdkBuildProjectRole` is generated by the `SimpleSynthAction`. If you are rolling your own action and overriding the role, you don't have to worry about it. Are there others blocking you?

You can also override the `owner` in `ActionProperties`. Make sure you set it to `"AWS"`.

**Comment by peterwoodworth** [comment]
@ColeSiegelTR have we been able to help you?

**Comment by peterwoodworth** [comment]
Please ping me if you want to reopen this issue! If this helped you, feedback would be appreciated 😄

**Comment by EysaN** [comment]
@peterwoodworth @ColeSiegelTR 

After 2 years, still I have not found a good documentation how to achieve this in CDK python. I would appreciate any feedback.

I have the following pipeline:

```python
from aws_cdk import Stack
from aws_cdk.pipelines import CodePipeline, CodePipelineSource, ShellStep, CodeBuildOptions


class MyPipeline(Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, 'appname-' + construct_id + '-pipeline', env=kwargs.get('env'))
        cdk_pipeline_role = iam.Role.from_role_arn(
            self, "cdk_pipeline_role",
            role_arn='role_arn',
            mutable=False
        )
        
        artifact_bucket = s3.Bucket.from_bucket_name(self, "cdk-appname", "cdk-appname")
        
        self.pipeline = CodePipeline(
            self, "appname-pipeline",
            artifact_bucket=artifact_bucket,
            pipeline_name="appname-pipeline",
            publish_assets_in_parallel=False,
            role=cdk_pipeline_role,
            self_mutation=False,
            code_build_defaults=CodeBuildOptions(),  # I tried configuring the CodeBuild project here but it made no difference
            synth=ShellStep(
                "Synth",
                input=CodePipelineSource.s3(
                    bucket=artifact_bucket,
                    object_key="appname",
                    role=cdk_pipeline_role
                ),
                install_commands=[
                    "npm install -g aws-cdk",
                    "python3 -m pip install -r requirements.txt",
                    "python3 -m pip install -r requirements-dev.txt"
                ],
                commands=[
                    "cdk synth"
                ],
            )
        )
```

And it is creating three IAM roles:
1. appnamepipelinePipelineBuildSynthCdkBuildProjectRole0660F84A
2. appnamepipelineCodeBuildActionRoleE18FF734
3. appnamepipelineAssetsFileRole7B3A6916

For compliance and security reasons, we should use existing roles instead, but I cannot wrap my head around it to find where those roles should fit in this pipeline above.

**Comment by HFR1994** [comment]
I got it working but now how I expected. Here is my implementation:

```
import {Artifact, Pipeline, PipelineType} from "aws-cdk-lib/aws-codepipeline";
import {S3SourceAction, S3Trigger} from "aws-cdk-lib/aws-codepipeline-actions";
import {
    AccountPrincipal,
    Effect,
    Policy,
    PolicyDocument,
    PolicyStatement,
    Role,
    ServicePrincipal
} from "aws-cdk-lib/aws-iam";
import {Stack} from "aws-cdk-lib";
import {Bucket} from "aws-cdk-lib/aws-s3";

const sourceArtifact = new Artifact();
const artifactsBucket = Bucket.fromBucketArn(this, "ArtifactsBucket", "arn:aws:s3:::my-fake-bucket");

const codePipelineRole = new Role(this, 'CodePipelineRole', {
    assumedBy: new ServicePrincipal("codepipeline.amazonaws.com"),
    description: 'CodePipeline rol shared across portafolio',
    roleName: "codepipeline-deployment-role"
});

const s3CodepipelineActionRole = new Role(this, "s3CodepipelineActionRole", {
    assumedBy: new AccountPrincipal(Stack.of(this).account),
    roleName: "s3-action-role",
    inlinePolicies: {
        "S3Object": new PolicyDocument({
            statements: [
                new PolicyStatement({
                    actions: [
                        "s3:GetBucket*",
                        "s3:GetObject*",
                        "s3:List*"
                    ],
                    resources: [
                        artifactsBucket.arnForObjects("deployments/file.zip")
                    ],
                })
            ]
        }),
        "S3Bucket": new PolicyDocument({
            statements: [
                new PolicyStatement({
                    actions: [
                        "s3:Abort*",
                        "s3:DeleteObject*",
                        "s3:PutObject",
                        "s3:PutObjectLegalHold",
                        "s3:PutObjectRetention",
                        "s3:PutObjectTagging",
                        "s3:PutObjectVersionTagging"
                    ],
                    resources: [
                        artifactsBucket.arnForObjects("*"),
                        artifactsBucket.bucketArn
                    ],
                })
            ]
        }),
    }
});

codePipelineRole.attachInlinePolicy(
    new Policy(this, 'StandardCodePipelineRules', {
        statements: [
            new PolicyStatement({
                effect: Effect.ALLOW,
                actions: [
                    "sts:AssumeRole"
                ],
                resources: [
                    codePipelineRole.s3CodepipelineActionRole.roleArn
                ]
            })
        ]
    })
)


const pipeline = new Pipeline(this, "Pipeline", {
    pipelineName: "mi-pipeline",
    crossAccountKeys: false,
    restartExecutionOnUpdate: true,
    pipelineType: PipelineType.V2,
    role: codePipelineRole,
    stages: [
        {
            stageName: "Source",
            actions: [
                new S3SourceAction({
                    actionName: "S3Artifact",
                    output: sourceArtifact,
                    bucket: artifactsBucket,
                    bucketKey: `deployments/file.zip`,
                    trigger: S3Trigger.EVENTS,
                    role: s3CodepipelineActionRole
                })
            ]
        } ,
        // Keep adding stages
    ]
});
```

As you may notice the `s3CodepipelineActionRole` defines a trust policy to `new AccountPrincipal(Stack.of(this).account)` I really don't understand why but it's the only way you can get it to work, if you define the trust policy directly to the CodePipeline Role (codePipelineRole in my case) it will throw the following message

> Resource handler returned message: "arn:aws:iam::1234567890:role/codepipeline-deployment-role is not authorized to perform AssumeRole on role arn:aws:iam::607274711708:role/s3-action-role (Service: CodePipeline, Status Code: 400, Request ID: <<SOME_ID>>)" (RequestToken: <<SOME_TOKEN>>, HandlerErrorCode: GeneralServiceException

Basically to solve your question, every aws-codepipeline-actions action has it's own parameter to specify a role regardless of it's origin (CodeBuild, S3 etc)...

Ej: 

```
import {
    CodeBuildAction,
} from "aws-cdk-lib/aws-codepipeline-actions";

const sourceArtifact = new Artifact();
const templateArtifact = new Artifact();

new CodeBuildAction({
            actionName: 'SomeCodeBuildAction',
            project: this.createRoleInDeployment,  <---- CodeBuild Project
            role: this.auditCodepipelineActionRole, <---- Role parameter for CodeBuild
            input: sourceArtifact,
            outputs: [templateArtifact],
            runOrder: 1,
        })
```

**Comment by gmournos** [comment]
Hi,

Resources sprawl is a problem with the CDK pipeline libraries. This is esp. true for roles, where every pipeline instance creates around 9 new roles.  
I addressed these and other such problems in a pipeline libary I created:

https://medium.com/@gmournos/creating-professional-aws-native-delivery-pipelines-with-cdk-c2aed59be6e7
https://medium.com/@gmournos/creating-professional-aws-native-delivery-pipelines-with-cdk-part-2-68fde253c34e

It is quite robust, I have been using it in production for one year.
I built the Medium articles it as a tutorial, so if you do not want to use it as is, you can easily pick the individual steps that treat resources sprawl and adjust them in your project.

Specifically:
For the roles, I used a Cloudformation macro to do the disassociation from the pipeline and the deletion of the extra roles and policies that the libraries create.
For keys and buckets reuse in cross region deployments, I created a custom resource that finds the cross-region KMS keys, so that they can be looked up from the pipeline stack.
For pipelines/codebuild projects, I added utilities that lookup "old" pipelines and delete their corresponding Cloudformation stacks.

Hope this helps

G Mournos
